app/scripts/migrations/107.ts

```diff
@@ -0,0 +1,82 @@
+import { cloneDeep } from 'lodash';
+import { hasProperty, isObject } from '@metamask/utils';
+
+export const version = 107;
+
+interface Account {
+  address: string;
+  balance: string;
+}
+
+interface AccountTrackerControllerState {
+  accountsByChainId: Record<string, Record<string, Account>>;
+}
+
+/**
+ * Migrates state from the now removed CachedBalancesController to the AccountTrackerController and formats it accordingly.
+ *
+ * @param originalVersionedData - Versioned MetaMask extension state, exactly what we persist to dist.
+ * @param originalVersionedData.meta - State metadata.
+ * @param originalVersionedData.meta.version - The current state version.
+ * @param originalVersionedData.data - The persisted MetaMask state, keyed by controller.
+ * @returns Updated versioned MetaMask extension state.
+ */
+export async function migrate(originalVersionedData: {
+  meta: { version: number };
+  data: Record<string, unknown>;
+}) {
+  const versionedData = cloneDeep(originalVersionedData);
+  versionedData.meta.version = version;
+  versionedData.data = transformState(versionedData.data);
+  return versionedData;
+}
+
+function transformState(state: Record<string, unknown>) {
+  if (
+    !hasProperty(state, 'CachedBalancesController') ||
+    !isObject(state.CachedBalancesController) ||
+    !hasProperty(state.CachedBalancesController, 'cachedBalances') ||
+    !isObject(state.CachedBalancesController.cachedBalances) ||
+    !hasProperty(state, 'AccountTracker')
+  ) {
+    return state;
+  }
+
+  const accountTrackerControllerState = (state.AccountTracker ||
+    {}) as AccountTrackerControllerState;
```

Comment:
if `state.AccountTracker` isn't an object for some reason, would you want to just abort instead?

Reply:
I guess you wouldn't. But that also means we should probably remove `!hasProperty(state, 'AccountTracker')` above then

Reply:
yes probably makes sense!

Reply:
done here: https://github.com/MetaMask/metamask-extension/pull/22325/commits/6f75e5ca550026843926b2e8bf85e1d58d6a3df7

Reply:
done here: https://github.com/MetaMask/metamask-extension/pull/22325/commits/d2ae66e36a1a7b918dff68f1a5ae4ec20a8603c1